File: src/bot/cogs/brands.py
"""Brand information commands cog."""

import discord
from discord.ext import commands

from models import get_db
from services.database import BrandService, UserService
from utils.logger import get_logger

logger = get_logger(__name__)


class BrandsCommands(commands.Cog):
    """Commands for brand information."""

    def __init__(self, bot: commands.Bot):
        self.bot = bot

    @commands.command(name="brands", aliases=["brand"])
    async def brand_info(self, ctx: commands.Context, *, brand_name: str = None):
        """
        Get information about sustainable brands or search for a specific brand.

        Usage: !brands [brand_name]
        Example: !brands patagonia
        """
        async with ctx.typing():
            db = next(get_db())
            try:
                # Track user
                user = UserService.get_or_create(db, ctx.author.id, ctx.author.name)
                user.add_points(5)
                UserService.update(db, user)

                if brand_name:
                    # Search for specific brand
                    brand = BrandService.get_by_name(db, brand_name)

                    if not brand:
                        await ctx.send(f"❌ Brand '{brand_name}' not found.")
                        return

                    # Create detailed embed
                    embed = discord.Embed(
                        title=f"🏷️ {brand.name}",
                        description=brand.description or "No description available",
                        color=discord.Color.green() if brand.overall_rating >= 60 else discord.Color.orange(),
                        url=brand.website if brand.website else None
                    )

                    embed.add_field(
                        name="Overall Rating",
                        value=f"**{brand.overall_rating:.1f}/100**\n{brand.get_rating_summary()}",
                        inline=False
                    )

                    # Ratings
                    embed.add_field(
                        name="🌍 Environmental",
                        value=f"{brand.environmental_rating:.1f}/100",
                        inline=True
                    )
                    embed.add_field(
                        name="👥 Labor",
                        value=f"{brand.labor_rating:.1f}/100",
                        inline=True
                    )
                    embed.add_field(
                        name="🐾 Animal Welfare",
                        value=f"{brand.animal_welfare_rating:.1f}/100",
                        inline=True
                    )

                    # Certifications
                    badges = brand.get_certification_badges()
                    if badges:
                        embed.add_field(
                            name="Certifications",
                            value="\n".join(f"✓ {badge}" for badge in badges),
                            inline=False
                        )

                    # Additional info
                    if brand.country:
                        embed.add_field(name="Country", value=brand.country, inline=True)
                    if brand.price_range:
                        embed.add_field(name="Price Range", value=brand.price_range, inline=True)
                    if brand.transparency_score:
                        embed.add_field(
                            name="Transparency",
                            value=f"{brand.transparency_score:.1f}/100",
                            inline=True
                        )

                    if brand.good_on_you_rating:
                        embed.add_field(
                            name="Good On You Rating",
                            value=brand.good_on_you_rating,
                            inline=False
                        )

                    embed.set_footer(text=f"Requested by {ctx.author.name} | +5 points")

                    await ctx.send(embed=embed)

                else:
                    # Show top-rated brands
                    brands = BrandService.get_top_rated(db, limit=10)

                    if not brands:
                        await ctx.send("❌ No brands found in database.")
                        return

                    embed = discord.Embed(
                        title="🌟 Top Sustainable Brands",
                        description="Here are the highest-rated sustainable fashion brands:",
                        color=discord.Color.gold()
                    )

                    for i, brand in enumerate(brands, 1):
                        embed.add_field(
                            name=f"{i}. {brand.name} - {brand.overall_rating:.1f}/100",
                            value=brand.get_rating_summary(),
                            inline=False
                        )

                    embed.set_footer(text=f"Use !brands <name> for detailed info | +5 points")

                    await ctx.send(embed=embed)

            except Exception as e:
                logger.error(f"Error in brands command: {e}", exc_info=True)
                await ctx.send("❌ An error occurred while fetching brand data.")
            finally:
                db.close()


async def setup(bot: commands.Bot):
    """Set up the cog."""
    await bot.add_cog(BrandsCommands(bot))

File: src/commands/brands.rs
// SPDX-License-Identifier: PMPL-1.0-or-later
//! Brands cog: `brands`.

use poise::serenity_prelude as serenity;

use super::{colour, say, send_embed};
use crate::services::{BrandService, UserService};
use crate::{typing, Context, Error};

/// Get information about sustainable brands, or detail for one brand.
#[poise::command(prefix_command, aliases("brand"))]
pub async fn brands(
    ctx: Context<'_>,
    #[rest]
    #[description = "Brand name (optional)"]
    brand_name: Option<String>,
) -> Result<(), Error> {
    typing(&ctx).await;
    let db = &ctx.data().db;

    let mut user =
        UserService::get_or_create(db, ctx.author().id.get() as i64, &ctx.author().name).await?;
    user.add_points(5);
    UserService::update(db, &user).await?;

    match brand_name {
        Some(name) => {
            let brand = match BrandService::get_by_name(db, &name).await? {
                Some(b) => b,
                None => return say(&ctx, format!("❌ Brand '{name}' not found.")).await,
            };

            let mut embed = serenity::CreateEmbed::new()
                .title(format!("🏷️ {}", brand.name))
                .description(
                    brand
                        .description
                        .clone()
                        .unwrap_or_else(|| "No description available".into()),
                )
                .colour(if brand.overall_rating >= 60.0 {
                    colour::green()
                } else {
                    colour::orange()
                });
            if let Some(url) = brand.website.clone() {
                embed = embed.url(url);
            }
            embed = embed
                .field(
                    "Overall Rating",
                    format!(
                        "**{:.1}/100**\n{}",
                        brand.overall_rating,
                        brand.get_rating_summary()
                    ),
                    false,
                )
                .field(
                    "🌍 Environmental",
                    format!("{:.1}/100", brand.environmental_rating),
                    true,
                )
                .field("👥 Labor", format!("{:.1}/100", brand.labor_rating), true)
                .field(
                    "🐾 Animal Welfare",
                    format!("{:.1}/100", brand.animal_welfare_rating),
                    true,
                );

            let badges = brand.get_certification_badges();
            if !badges.is_empty() {
                embed = embed.field(
                    "Certifications",
                    badges
                        .iter()
                        .map(|b| format!("✓ {b}"))
                        .collect::<Vec<_>>()
                        .join("\n"),
                    false,
                );
            }
            if let Some(c) = &brand.country {
                embed = embed.field("Country", c, true);
            }
            if let Some(p) = &brand.price_range {
                embed = embed.field("Price Range", p, true);
            }
            if let Some(t) = brand.transparency_score {
                embed = embed.field("Transparency", format!("{t:.1}/100"), true);
            }
            if let Some(g) = &brand.good_on_you_rating {
                embed = embed.field("Good On You Rating", g, false);
            }
            embed = embed.footer(serenity::CreateEmbedFooter::new(format!(
                "Requested by {} | +5 points",
                ctx.author().name
            )));
            send_embed(&ctx, embed).await
        }
        None => {
            let brands = BrandService::get_top_rated(db, 10).await?;
            if brands.is_empty() {
                return say(&ctx, "❌ No brands found in database.").await;
            }
            let mut embed = serenity::CreateEmbed::new()
                .title("🌟 Top Sustainable Brands")
                .description("Here are the highest-rated sustainable fashion brands:")
                .colour(colour::gold());
            for (i, b) in brands.iter().enumerate() {
                embed = embed.field(
                    format!("{}. {} - {:.1}/100", i + 1, b.name, b.overall_rating),
                    b.get_rating_summary(),
                    false,
                );
            }
            embed = embed.footer(serenity::CreateEmbedFooter::new(
                "Use !brands <name> for detailed info | +5 points",
            ));
            send_embed(&ctx, embed).await
        }
    }
}
